### ISSUE TITLE
Deprecate mutability of Settings objects

### ISSUE THREAD

**Issue by DRMacIver** [DRMacIver/hypothesis#199]
So Settings objects are currently mutable. It's even part of their recommended usage pattern. For example, from the documentation:

```pycon
>>> from hypothesis import Settings
>>> s = Settings()
>>> s.max_examples
200
>>> Settings.default.max_examples = 100
>>> t = Settings()
>>> t.max_examples
100
>>> s.max_examples
200
```

It's undocumented, but you can even assign Settings.default (historically this has not worked correctly, but it started working correctly in a recent version):

```
>>> from hypothesis import Settings
>>> Settings().max_examples
200
>>> Settings.default = Settings(max_examples=10)
>>> Settings().max_examples
10
```

I think this mostly leads to confusion, and it's particularly confusing in light of #154, in which @Bachmann1234 introduced a much better way to change settings globally.

My proposal is that all of these assignments should now start to emit deprecation warnings, with the recommendation of using the new profiles mechanism and/or local settings objects, with the goal of settings becoming immutable in Hypothesis 2.0.

Does anyone have any use cases they feel are not well represented by this? Votes aye or nay?

**Comment by kxepal** [comment]
+1

**Comment by Bachmann1234** [comment]
aye! (github needs a +1 button)

**Comment by DRMacIver** [comment]
Right, the general response to this suggestion has basically been unanimous enthusiasm, so I'm going to go with this. :-)

**Comment by DRMacIver** [comment]
This is now in a released version as of 1.12.0
